source/extensions/quic_listeners/quiche/envoy_quic_client_session.cc

```diff
@@ -103,5 +103,21 @@ void EnvoyQuicClientSession::OnTlsHandshakeComplete() {
   raiseConnectionEvent(Network::ConnectionEvent::Connected);
 }
 
+size_t EnvoyQuicClientSession::WriteHeadersOnHeadersStream(
+    quic::QuicStreamId id, spdy::SpdyHeaderBlock headers, bool fin,
+    const spdy::SpdyStreamPrecedence& precedence,
+    quic::QuicReferenceCountedPointer<quic::QuicAckListenerInterface> ack_listener) {
+  ASSERT(!quic::VersionUsesHttp3(transport_version()));
+  // gQUIC headers are sent on a dedicated stream. Only count the bytes sent against
+  // connection level watermark buffer. Do not count them into stream level
+  // watermark buffer, because it is impossible to identify which byte belongs
+  // to which stream when the buffered bytes are drained in headers stream.
+  // This updater may be in the scope of another one in OnCanWrite(), in such
+  // case, this one doesn't update the watermark.
```

Comment:
To make sure I understand this,
is it the case that sometimes we get onCanWrite, and write stream data, at which point watermark tracking happens from the onCanWrite updater, and sometimes we end up writing due to upstream data being flushed, at which point this tracker applies?

Reply:
Yes, if this is in the scope of OnCanWrite(), the udpater in OnCanWrite() will track the buffered bytes in header stream.
In other cases, the udpater here will do the work.